apps/cmp.c

```diff
@@ -0,0 +1,3557 @@
+ * returns 1 on success, 0 on error.
+ */
+#define X509_STORE_EX_DATA_HOST 0
+static int truststore_set_host_etc(X509_STORE *ts, char *host)
+{
+    X509_VERIFY_PARAM *ts_vpm = X509_STORE_get0_param(ts);
+
+    /* first clear any host names, IP, and email addresses */
+    if (!X509_VERIFY_PARAM_set1_host(ts_vpm, NULL, 0)
+            || !X509_VERIFY_PARAM_set1_ip(ts_vpm, NULL, 0)
+            || !X509_VERIFY_PARAM_set1_email(ts_vpm, NULL, 0))
+        return 0;
+    X509_VERIFY_PARAM_set_hostflags(ts_vpm,
+                                    X509_CHECK_FLAG_ALWAYS_CHECK_SUBJECT |
+                                    X509_CHECK_FLAG_NO_PARTIAL_WILDCARDS);
+    /*
+     * Unfortunately there is no OpenSSL API function for retrieving the hosts/
+     * ip entries in X509_VERIFY_PARAM. So we store the host value in ex_data
+     * for use in cert_verify_cb() below.
+     */
+    if (!X509_STORE_set_ex_data(ts, X509_STORE_EX_DATA_HOST, host))
+        return 0;
+    return (host != NULL && X509_VERIFY_PARAM_set1_ip_asc(ts_vpm, host))
+        || X509_VERIFY_PARAM_set1_host(ts_vpm, host, 0);
+}
+
+static X509_STORE *sk_X509_to_store(X509_STORE *store /* may be NULL */,
+                                    const STACK_OF(X509) *certs /* may NULL */)
+{
+    int i;
+
+    if (store == NULL)
+        store = X509_STORE_new();
+    if (store == NULL)
+        return NULL;
+    for (i = 0; i < sk_X509_num(certs); i++) {
+        if (!X509_STORE_add_cert(store, sk_X509_value(certs, i))) {
+            X509_STORE_free(store);
+            return NULL;
+        }
+    }
+    return store;
+}
+
+/*-
+ * Writes OSSL_CMP_MSG DER-encoded to the file specified with outfile
+ *
+ * returns 1 on success, 0 on error
+ */
+static int write_PKIMESSAGE(OSSL_CMP_CTX *ctx,
+                            const OSSL_CMP_MSG *msg, char **filenames)
+{
+    char *file;
+    FILE *f;
+    int res = 0;
+
+    if (msg == NULL || filenames == NULL) {
+        CMP_err("NULL arg to write_PKIMESSAGE");
+        return 0;
+    }
+    if (*filenames == NULL) {
+        CMP_err("not enough file names have been provided for writing message");
+        return 0;
+    }
+
+    file = *filenames;
+    *filenames = next_item(file);
+    f = fopen(file, "wb");
+    if (f == NULL) {
+        CMP_err1("cannot open file '%s' for writing", file);
+    } else {
+        unsigned char *out = NULL;
+        int len = i2d_OSSL_CMP_MSG(msg, &out);
+
+        if (len >= 0) {
+            if ((size_t)len == fwrite(out, sizeof(*out), len, f))
+                res = 1;
+            else
+                CMP_err1("cannot write file '%s'", file);
+            OPENSSL_free(out);
+        }
+        fclose(f);
+    }
+    return res;
+}
+
+/*-
+ * Reads a DER-encoded OSSL_CMP_MSG from the file specified in infile
+ * The OSS_CMP_MSG must be freed by the caller
+ *
+ * returns a pointer to the parsed OSSL_CMP_MSG, null on error
+ */
+static OSSL_CMP_MSG *read_PKIMESSAGE(OSSL_CMP_CTX *ctx, char **filenames)
```

Comment:
what ctx is needed for?

Reply:
Good spotting - it's not use there any more, neither in `write_PKIMESSAGE()`; removed.